cardano-node/src/Cardano/CLI/Tx.hs

```diff
@@ -0,0 +1,114 @@
+import qualified Data.Map.Strict as Map
+import           Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as TL
+import qualified Data.Vector as V
+import qualified Formatting as F
+
+import           Cardano.Binary (reAnnotate)
+import           Cardano.Chain.Common
+import qualified Cardano.Chain.Common as CC.Common
+import           Cardano.Chain.Genesis as CC.Genesis
+import           Cardano.Chain.UTxO
+import qualified Cardano.Chain.UTxO as CC.UTxO
+import           Cardano.Crypto (SigningKey(..), ProtocolMagicId)
+import qualified Cardano.Crypto.Hashing as Crypto
+import qualified Cardano.Crypto.Signing as Crypto
+import           Ouroboros.Consensus.Ledger.Byron
+import           Ouroboros.Consensus.Ledger.Byron.Config
+
+
+prettyAddress :: CC.Common.Address -> Text
+prettyAddress addr = TL.toStrict
+  $  F.format CC.Common.addressF         addr <> "\n"
+  <> F.format CC.Common.addressDetailedF addr
+
+txSpendUTxOByronPBFT
+  :: CC.Genesis.Config
+  -> SigningKey
+  -> NonEmpty TxIn
+  -> NonEmpty TxOut
+  -> GenTx (ByronBlockOrEBB ByronConfig)
+txSpendUTxOByronPBFT gc sk ins outs =
+  let txattrs  = mkAttributes ()
+      tx       = UnsafeTx ins outs txattrs
+      wit      = signTxId (configProtocolMagicId gc) sk (Crypto.hash tx)
+      ATxAux atx awit =
+        mkTxAux tx . V.fromList . take (NE.length ins) $ repeat wit
+  in mkByronTx $ ATxAux (reAnnotate atx) (reAnnotate awit)
+
+txSpendGenesisUTxOByronPBFT
+  :: CC.Genesis.Config
+  -> SigningKey
+  -> Address
+  -> NonEmpty TxOut
+  -> GenTx (ByronBlockOrEBB ByronConfig)
+txSpendGenesisUTxOByronPBFT gc sk genAddr outs =
+  let txattrs  = mkAttributes ()
+      tx       = UnsafeTx (pure txIn) outs txattrs
+      txIn    :: CC.UTxO.TxIn
+      txIn     = genesisUTxOTxIn gc sk genAddr
+      wit      = signTxId (configProtocolMagicId gc) sk (Crypto.hash tx) 
+      ATxAux atx awit = mkTxAux tx . V.fromList . pure $ wit
+  in mkByronTx $ ATxAux (reAnnotate atx) (reAnnotate awit)
+
+signTxId :: ProtocolMagicId -> SigningKey -> TxId -> CC.UTxO.TxInWitness
+signTxId pmid sk txid = CC.UTxO.VKWitness
+  (Crypto.toVerification sk)
+  (Crypto.sign
+    pmid
+    Crypto.SignTx
+    sk
+    (CC.UTxO.TxSigData txid))
+
+-- | Given a genesis, and a pair of signing key and address, reconstruct a TxIn
+--   corresponding to the genesis UTxO entry.
+genesisUTxOTxIn :: CC.Genesis.Config -> SigningKey -> Address -> CC.UTxO.TxIn
+genesisUTxOTxIn gc genSk genAddr =
+  let vk         = Crypto.toVerification genSk
+      handleMissingAddr :: Maybe CC.UTxO.TxIn -> CC.UTxO.TxIn
+      handleMissingAddr  = fromMaybe . error
+        $  "\nGenesis UTxO has no address\n"
+        <> (T.unpack $ prettyAddress genAddr)
+        <> "\n\nIt has the following, though:\n\n"
+        <> Cardano.Prelude.concat (T.unpack . prettyAddress <$> Map.keys initialUtxo)
+
+      initialUtxo :: Map Address (CC.UTxO.TxIn, CC.UTxO.TxOut)
+      initialUtxo =
```

Comment:
There is a lot going on in this function. It may be a good idea to factor out these let bindings into separate functions with comments.

Reply:
This is a code motion PR.